py/kubeflow/testing/cleanup_ci.py

```diff
@@ -128,45 +135,110 @@ def cleanup_endpoints(args):
   logging.info("expired services:\n%s", "\n".join(expired))
 
 
-def cleanup_disks(args):
+def cleanup_endpoints(args):
   credentials = GoogleCredentials.get_application_default()
 
-  compute = discovery.build('compute', 'v1', credentials=credentials)
-  disks = compute.disks()
+  services_management = discovery.build('servicemanagement', 'v1', credentials=credentials)
+  services = services_management.services()
+  rollouts = services.rollouts()
   next_page_token = None
 
   expired = []
   unexpired = []
   unmatched = []
 
-  for zone in args.zones.split(","):
-    while True:
-      results = disks.list(project=args.project,
-                           zone=zone,
-                           pageToken=next_page_token).execute()
-      if not "items" in results:
-        break
-      for d in results["items"]:
-        name = d["name"]
-        if not is_match_disk(name):
-          unmatched.append(name)
-          continue
+  while True:
+    results = services.list(producerProjectId=args.project,
+                            pageToken=next_page_token).execute()
+
+    for s in results["services"]:
+      name = s["serviceName"]
+      if not is_match(name):
+        unmatched.append(name)
+        continue
 
-        age = getAge(d["creationTimestamp"])
+      all_rollouts = rollouts.list(serviceName=name).execute()
+      is_expired = False
+      if not all_rollouts.get("rollouts", []):
+        logging.info("Service %s has no rollouts", name)
+        is_expired = True
+      else:
+        r = all_rollouts["rollouts"][0]
+        create_time = date_parser.parse(r["createTime"])
+
+        now = datetime.datetime.now(create_time.tzinfo)
+
+        age = now - create_time
         if age > datetime.timedelta(hours=args.max_age_hours):
-          logging.info("Deleting disk: %s, age = %r", name, age)
-          response = disks.delete(project=args.project, zone=zone, disk=name)
+          is_expired = True
+
+      if is_expired:
+        logging.info("Deleting service: %s", name)
+        is_expired = True
+        if not args.dryrun:
+          services.delete(serviceName=name).execute()
+        expired.append(name)
+      else:
+        unexpired.append(name)
+
+    if not "nextPageToken" in results:
+      break
+    next_page_token = results["nextPageToken"]
+
+
+  logging.info("Unmatched services:\n%s", "\n".join(unmatched))
+  logging.info("Unexpired services:\n%s", "\n".join(unexpired))
+  logging.info("expired services:\n%s", "\n".join(expired))
+
+def cleanup_firewall_rules(args):
+  credentials = GoogleCredentials.get_application_default()
+
+  compute = discovery.build('compute', 'v1', credentials=credentials)
+  firewalls = compute.firewalls()
+  next_page_token = None
+
+  expired = []
+  unexpired = []
+  unmatched = []
+
+  while True:
```

Comment:
use retry?

Reply:
The while loop here is to handle pagination.

If we have retryable errors it would likely be on the list & delete methods. I'm not sure what errors to retry on though so my sense would be to wait until we see which errors are causing retryable failures and then deal with them.